### DISCUSSION TITLE
customdata not working anymore after update from 2.1.0 to 3.3.1

### DISCUSSION THREAD

**Discussion by MrSplit** [dbuezas/lovelace-plotly-graph-card#242] [Q&A] [answered]
I updated yesterday from 2.1.0 to 3.3.1 but the customdata variable seems not to be working anymore.
The following setting:
```
    hovertemplate: >-
          <b>%{customdata.name}</b><br><i>%{x}</i><br>
          %{y:.1f}%{customdata.unit_of_measurement} <extra></extra>
```
is giving me a result like this:
>     %{customdata.name}
>     Feb 6, 2023, 05:40
>     18,7%{customdata.unit_of_measurement}

Can anybody help?
Thanks!

**Comment by dbuezas** [comment, marked as answer]
Yes, that's a breaking change I didn't document because the default cases are covered by the new defaults. Sorry for that.
The reason behind the change is performance, this was using quite a bit of ram because the custom data needed to be replicated for each datapoint.

You have two options:
1. bring back the old data
```yaml
customdata: $ex ys.map(()=>meta)
hovertemplate: >-
          <b>%{customdata.friendly_name}</b><br><i>%{x}</i><br>
          %{y:.1f}%{customdata.unit_of_measurement} <extra></extra>
```

2. Or efficiently: (this is better)
```yaml
hovertemplate: |
    $ex `<b>${get(".name")}</b><br><i>%{x}</i><br>
    %{y:.1f} ${get(".unit_of_measurement")}
    <extra></extra>`
```
`get(".name")` means "look at the name attribute in the same entity" (and same depth in the yaml). The name is set from the entity's friendly_name (from Home Assistant) or whatever you may have overridden it with via yaml.
This avoids having to generate more big arrays of repeated data. It works by setting the name and unit directly in the template as text (not as template variables that need to be found inside the right index of the customdata array). 

You can see how all the other defaults are defined [here](https://github.com/dbuezas/lovelace-plotly-graph-card/blob/master/src/parse-config/defaults.ts)

**Reply by MrSplit** [reply]
Thanks David for the quick reply.
I tried the option 2, but I get the following error:
_Error: at [entities.0.hovertemplate]: Since [entities.0.unit_of_measurement] is a $fn, it has to be defined before [entities.0.hovertemplate]_
Any idea?
Thanks!

**Reply by dbuezas** [reply]
Oh, right that default, also function and as a default it goes at the end. I need to add the concept of pure functions. In the meanwhile, you can also define unit_of_measurement in the same manner (look at the defaults file) and put it before text template you'll be fine

**Reply by dbuezas** [reply]
Actually, you can just do:
```yaml
unit_of_measurement: $ex meta.unit_of_measurement
```

**Reply by dbuezas** [reply]
(make sure it is defined before texttemplate)

**Reply by MrSplit** [reply]
This worked, thanks!

**Reply by dbuezas** [reply]
I put an issue to get rid of this weirdness: #246
